src/net/ifaddrs/mod.rs

```diff
@@ -0,0 +1,273 @@
+        }
+
+        m
+    }
+}
+
+impl<'a> Drop for InterfaceAddrs<'a> {
+    fn drop(&mut self) {
+        if self.do_free {
+            // UNSAFETY: Calling libc FFI function which frees previously allocated
+            // memory.
+            unsafe {
+                // Ask the libc to drop free the memory it allocated when
+                // the struct was created.
+                libc::freeifaddrs(self.inner as *mut libc::ifaddrs);
+            }
+        }
+    }
+}
+
+
+/// Represents the configuration and state of a network interface.
+/// Interfaces are uniquely identified by name, and each interface is likely
+/// to be referred to multiple times, e.g. one for IPv4 and one for IPv6.
+#[derive(Debug, Clone)]
+pub struct InterfaceAddr<'a> {
+    /// The name of the interface
+    pub name: String,
+
+    /// The address assigned to the interface for this protocol.
+    /// A value of `None` means the libc reported a type of address that
+    /// `std::net` doesn't understand.
+    pub address: Option<IfAddrValue<'a>>,
+
+    /// The netmasks assigned to the interface for this protocol.
+    /// A value of `None` means the libc reported a type of address that
+    /// `std::net` doesn't understand.
+    pub netmask: Option<IfAddrValue<'a>>,
+
+    /// The ifu assigned to the interface for this protocol.
+    /// A value of `{Broadcast, Destination}Addr(None)` means the libc reported
+    /// a type of address that `std::net` doesn't understand, while a value of
+    /// `Neither` means that the interface has neither a valid broadcast address
+    /// nor a point-to-point destination address.
+    pub ifu: InterfaceIfu<'a>,
+
+    /// Flags regarding the interface's behaviour and state
+    pub flags: IffFlags,
+}
+
+/// Represents the ifu of an interface: either its broadcast address or
+/// point-to-point destination address.
+#[derive(Debug, Clone)]
+pub enum InterfaceIfu<'a> {
+    BroadcastAddr(Option<IfAddrValue<'a>>),
+    DestinationAddr(Option<IfAddrValue<'a>>),
+    Neither,
+}
+
+
+impl<'a> Iterator for InterfaceAddrs<'a> {
+    type Item = InterfaceAddr<'a>;
+    fn next(&mut self) -> Option<InterfaceAddr<'a>> {
+        // If the current ifaddrs is None, there are no more ifaddrs to inspect
+        if self.current.is_none() {
+            return None;
+        }
+
+        // Workaround for the borrow checker being overzealous
+        // (without ptr_temp, self.current would technically
+        // "still be in use" when the loop ends, meaning we
+        // couldn't advance to the next struct)
+        let ptr_temp = self.current.clone();
+        let p = ptr_temp.as_ref().unwrap();
+
+        // Get a pointer to the interface's name
+        let name_ptr = p.ifa_name;
+        // Check that name_ptr isn't null.
+        if name_ptr.is_null() {
+            panic!("getifaddrs() gave an ifaddrs struct with a null ifa_name");
+        }
+
+        // UNSAFETY: Constructing CStr from pointer. If this pointer is
```

Comment:
Since the pointer being null is checked above with a panic if it's NULL at this point, it's a compiler error or something's super wrong with the CPU. I think this comment can be cleaned up a big to clarify that. Same for the one below.